codex/slotbuilder/slotbuilder.nim

```diff
@@ -0,0 +1,92 @@
+import std/math
+import std/sequtils
+import pkg/libp2p
+import pkg/chronos
+import pkg/chronicles
+import pkg/questionable/results
+import ../merkletree
+import ../stores
+import ../manifest
+import ../utils
+
+let
+  # TODO: Unified with the CellSize specified in branch "data-sampler"
+  # Number of bytes in a cell. A cell is the smallest unit of data used
+  # in the proving circuit.
+  CellSize* = 2048
+
+type
+  SlotBuilder* = object of RootObj
+    blockStore: BlockStore
+    manifest: Manifest
+    numberOfSlotBlocks: int
+
+proc new*(
+    T: type SlotBuilder,
+    blockStore: BlockStore,
+    manifest: Manifest
+): ?!SlotBuilder =
+
+  if not manifest.protected:
+    return failure("Can only create SlotBuilder using protected manifests.")
+
+  if (manifest.blocksCount mod manifest.ecK) != 0:
+    return failure("Number of blocks must be divisable by number of slots.")
+
+  if (manifest.blockSize.int mod CellSize) != 0:
+    return failure("Block size must be divisable by cell size.")
+
+  let numberOfSlotBlocks = manifest.blocksCount div manifest.ecK
+  success(SlotBuilder(
+    blockStore: blockStore,
+    manifest: manifest,
+    numberOfSlotBlocks: numberOfSlotBlocks
+  ))
+
+proc cellsPerBlock(self: SlotBuilder): int =
+  self.manifest.blockSize.int div CellSize
+
+proc selectSlotBlocks*(self: SlotBuilder, datasetSlotIndex: int): Future[?!seq[Cid]] {.async.} =
+  var cids = newSeq[Cid]()
+  let
+    datasetTreeCid = self.manifest.treeCid
+    blockCount = self.manifest.blocksCount
+    numberOfSlots = self.manifest.ecK
+    strategy = SteppedIndexingStrategy.new(0, blockCount - 1, numberOfSlots)
+
+  for datasetBlockIndex in strategy.getIndicies(datasetSlotIndex):
+    without slotBlockCid =? await self.blockStore.getCid(datasetTreeCid, datasetBlockIndex), err:
```

Comment:
One more comment thought from me:
 
As far as I understand, the purpose of this change is to build a tree for slot data. Currently you're using regular merkle tree (sha256), but the idea is to plug in poseidon2 tree whenever the implementation is ready, right?

If that's the case, we won't ever use the `getCid` function, we will use `getBlock(datasetTreeCid, index)` instead. So is it actually worth changing the interface of the `Blockstore` by adding yet another method (`getCid`) that soon will be rendered unused?

Reply:
Yeah that's a good point, and the reason why I'm not eager to merge this one yet. I want to swap out the merkle tree as this code is currently using it for the correct poseidon one first. That's still in progress. So this branch, and also the data-sampler, will have to float around for a while until that's settled.